### PR TITLE
Add Capabilties to user

### PR DESCRIPTION
**Pull request by Honey14**
https://app.clubhouse.io/simpledotorg/story/880/add-a-check-if-the-current-mo-is-allowed-to-log-a-teleconsult-record-at-the-current-facility

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/user/User.kt [hunk 1]
```diff
@@ -138,6 +143,53 @@ data class User(
     UNAUTHORIZED;
 
     class RoomTypeConverter : RoomEnumTypeConverter<LoggedInStatus>(LoggedInStatus::class.java)
+
+  }
+
+  @Parcelize
+  data class Capabilities(val canTeleconsult: CapabilityType) : Parcelable
+
+  sealed class CapabilityType : Parcelable {
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/user/User.kt:162]: Can we change the name to be more appropriate of the use case, please? 

`CapabilityType` seems like it will have different capabilities, but in fact, we are just setting the enable/disable status of the capabilities. WDYT?

File: app/src/main/java/org/simple/clinic/user/User.kt [hunk 2]
```diff
@@ -138,6 +143,53 @@ data class User(
...
+      @TypeConverter
+      fun fromEnum(enum: CapabilityType?) = TypeAdapter.fromEnum(enum)
+    }
+
+    class MoshiTypeAdapter {
+      @FromJson
+      fun toEnum(value: String?) = TypeAdapter.toEnum(value)
+
+      @ToJson
+      fun fromEnum(enum: CapabilityType) = TypeAdapter.fromEnum(enum)
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/user/User.kt:196]: Let's handle null cases as well here.

File: app/src/main/java/org/simple/clinic/user/User.kt [hunk 3]
```diff
@@ -138,6 +143,53 @@ data class User(
...
+    }
+
+    companion object {
+      @VisibleForTesting
+      fun random() = CapabilityType.TypeAdapter.knownMappings.keys.shuffled().first()
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/user/User.kt:201]: I don't think we need it since we are not using it anywhere.

File: app/src/main/java/org/simple/clinic/storage/migrations/Migration_73.kt [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package org.simple.clinic.storage.migrations
+
+import androidx.room.migration.Migration
+import androidx.sqlite.db.SupportSQLiteDatabase
+import org.simple.clinic.storage.inTransaction
+import javax.inject.Inject
+
+@Suppress("Classname")
+class Migration_73 @Inject constructor() : Migration(72, 73) {
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/storage/migrations/Migration_73.kt:9]: Just an FYI, since we are squshing the commits when merging you can do the migration in a different commit so that it will be easier to review.

File: app/src/androidTest/java/org/simple/clinic/storage/migrations/Migration73AndroidTest.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
+package org.simple.clinic.storage.migrations
+
+import org.junit.Test
+import org.simple.clinic.assertColumns
+
+class Migration73AndroidTest : BaseDatabaseMigrationTest(72, 73) {
+
+  @Test
+  fun `capabilities_should_be_added_to_the_user_table`() {
```
Comment (msasikanth) [inline app/src/androidTest/java/org/simple/clinic/storage/migrations/Migration73AndroidTest.kt:9]: Can we also add a test to make sure the `LoggedInUser ` data is retained after the migration?

File: app/src/main/java/org/simple/clinic/AppDatabase.kt [hunk 6]
```diff
@@ -89,7 +89,8 @@ import org.simple.clinic.util.room.UuidRoomTypeConverter
     PatientStatus.RoomTypeConverter::class,
     ReminderConsent.RoomTypeConverter::class,
     BloodSugarMeasurementType.RoomTypeConverter::class,
-    DeletedReason.RoomTypeConverter::class
+    DeletedReason.RoomTypeConverter::class,
+    User.CapabilityType.TypeAdapter::class
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/AppDatabase.kt]: ~~We need to add `RoomTypeConverter` here.~~ NVM it has been added in next commit

File: app/src/main/java/org/simple/clinic/login/LoginUserWithOtp.kt [hunk 7]
```diff
@@ -61,7 +61,8 @@ class LoginUserWithOtp @Inject constructor(
           loggedInStatus = User.LoggedInStatus.LOGGED_IN,
           registrationFacilityUuid = payload.registrationFacilityId,
           currentFacilityUuid = payload.registrationFacilityId,
-          teleconsultPhoneNumber = payload.teleconsultPhoneNumber
+          teleconsultPhoneNumber = payload.teleconsultPhoneNumber,
+          capabilities = null
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/login/LoginUserWithOtp.kt:65]: We need to pass the capabilities from the payload.

File: app/src/main/java/org/simple/clinic/login/pin/LoginPinEffectHandler.kt [hunk 8]
```diff
@@ -89,7 +89,8 @@ class LoginPinEffectHandler @AssistedInject constructor(
         loggedInStatus = User.LoggedInStatus.OTP_REQUESTED,
         registrationFacilityUuid = entry.registrationFacilityUuid!!,
         currentFacilityUuid = entry.registrationFacilityUuid,
-        teleconsultPhoneNumber = entry.teleconsultPhoneNumber
+        teleconsultPhoneNumber = entry.teleconsultPhoneNumber,
+        capabilities = null
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/login/pin/LoginPinEffectHandler.kt:93]: Setting `null` here is wrong, we need to update `OngoingLoginEntry` to have capabilities as well. Please refer `UsersApi#activate` API endpoint which will return `LoggedInUserPayload` which will have the capabilities.

File: app/src/main/java/org/simple/clinic/storage/migrations/Migration_73.kt [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  override fun migrate(database: SupportSQLiteDatabase) {
+    database.inTransaction {
+      execSQL("""
+        ALTER TABLE "LoggedInUser" ADD COLUMN "capability_canTeleconsult" TEXT
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/storage/migrations/Migration_73.kt:13]: Let's set the default value to `null`

File: app/src/main/java/org/simple/clinic/user/UserSession.kt [hunk 10]
```diff
@@ -92,7 +93,8 @@ class UserSession @Inject constructor(
           loggedInStatus = status,
           registrationFacilityUuid = payload.registrationFacilityId,
           currentFacilityUuid = payload.registrationFacilityId,
-          teleconsultPhoneNumber = payload.teleconsultPhoneNumber
+          teleconsultPhoneNumber = payload.teleconsultPhoneNumber,
+          capabilities = null
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/user/UserSession.kt:97]: we need to set capabilities from the payload here

File: app/src/main/java/org/simple/clinic/user/refreshuser/RefreshCurrentUser.kt [hunk 11]
```diff
@@ -81,7 +81,8 @@ class RefreshCurrentUser @Inject constructor(
           loggedInStatus = loggedInStatus,
           registrationFacilityUuid = payload.registrationFacilityId,
           currentFacilityUuid = storedUser.currentFacilityUuid,
-          teleconsultPhoneNumber = payload.teleconsultPhoneNumber
+          teleconsultPhoneNumber = payload.teleconsultPhoneNumber,
+          capabilities = null
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/user/refreshuser/RefreshCurrentUser.kt:85]: we need to set capabilities from the payload here

File: app/src/main/java/org/simple/clinic/user/registeruser/RegisterUser.kt [hunk 12]
```diff
@@ -71,7 +71,8 @@ class RegisterUser @Inject constructor(
           status = status,
           createdAt = createdAt,
           updatedAt = updatedAt,
-          teleconsultPhoneNumber = user.teleconsultPhoneNumber
+          teleconsultPhoneNumber = user.teleconsultPhoneNumber,
+          capabilities = null
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/user/registeruser/RegisterUser.kt:75]: we need to set capabilities from the user here

File: app/src/main/java/org/simple/clinic/user/registeruser/RegisterUser.kt [hunk 13]
```diff
@@ -100,7 +101,8 @@ class RegisterUser @Inject constructor(
           loggedInStatus = LOGGED_IN,
           registrationFacilityUuid = payload.registrationFacilityId,
           currentFacilityUuid = payload.registrationFacilityId,
-          teleconsultPhoneNumber = payload.teleconsultPhoneNumber
+          teleconsultPhoneNumber = payload.teleconsultPhoneNumber,
+          capabilities = null
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/user/registeruser/RegisterUser.kt:105]: we need to set capabilities from the payload here

File: app/src/main/java/org/simple/clinic/user/resetpin/ResetUserPin.kt [hunk 14]
```diff
@@ -72,7 +72,8 @@ class ResetUserPin @Inject constructor(
           loggedInStatus = RESET_PIN_REQUESTED,
           registrationFacilityUuid = registrationFacilityId,
           currentFacilityUuid = registrationFacilityId,
-          teleconsultPhoneNumber = teleconsultPhoneNumber
+          teleconsultPhoneNumber = teleconsultPhoneNumber,
+          capabilities = null
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/user/resetpin/ResetUserPin.kt:76]: we need to set capabilities from the payload here

**Review by msasikanth — CHANGES_REQUESTED** [review]
Overall PR looks good, left a few comments PTAL.

File: app/src/main/java/org/simple/clinic/user/User.kt (hunk 1 above)
Comment (Honey14) [reply app/src/main/java/org/simple/clinic/user/User.kt:162]: Should we name it as `Capability`?

Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/user/User.kt:162]: Probably something like `CapabilityStatus`?

File: app/src/main/java/org/simple/clinic/AppDatabase.kt [hunk 15]
```diff
@@ -100,7 +100,8 @@ import org.simple.clinic.util.room.UuidRoomTypeConverter
     DeletedReason.RoomTypeConverter::class,
     MedicineFrequency.RoomTypeConverter::class,
     TeleconsultationType.RoomTypeConverter::class,
-    RoomTypeConverter::class
+    RoomTypeConverter::class,
+    User.CapabilityStatus.TypeAdapter::class
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/AppDatabase.kt]: ~~You have to pass the `User. CapabilityStatus.RoomTypeConverter::class` here~~ NVM it has been done in next commit

File: app/src/main/java/org/simple/clinic/user/User.kt [hunk 16]
```diff
@@ -138,6 +142,48 @@ data class User(
     UNAUTHORIZED;
 
     class RoomTypeConverter : RoomEnumTypeConverter<LoggedInStatus>(LoggedInStatus::class.java)
+
+  }
+
+  @Parcelize
+  data class Capabilities(val canTeleconsult: CapabilityStatus) : Parcelable
+
+  sealed class CapabilityStatus : Parcelable {
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/user/User.kt:161]: Let's override `toString` for the sealed classes types since we will be using it as an enum.

File: app/src/androidTest/java/org/simple/clinic/storage/migrations/Migration75AndroidTest.kt [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+          "fullName" to "Joshua Dawson",
+          "phoneNumber" to "999555888",
+          "pinDigest" to "1111",
+          "status" to "WaitingForApproval",
+          "createdAt" to Instant.parse("2018-01-01T00:00:00Z"),
+          "updatedAt" to Instant.parse("2018-01-01T00:00:00Z"),
+          "loggedInStatus" to "LOGGED_IN",
+          "registrationFacilityUuid" to facilityUuid,
+          "currentFacilityUuid" to facilityUuid,
+          "teleconsultPhoneNumber" to "85459839458",
```
Comment (msasikanth) [inline app/src/androidTest/java/org/simple/clinic/storage/migrations/Migration75AndroidTest.kt:99]: Let's not use actual phone numbers in our code. Let's go with `1111111111` or something like that.

File: app/src/sharedTest/java/org/simple/clinic/TestData.kt [hunk 18]
```diff
@@ -1071,7 +1071,9 @@ object TestData {
       registrationFacilityUuid = registrationFacilityUuid,
       status = status,
       createdAt = createdAt,
-      updatedAt = updatedAt
+      updatedAt = updatedAt,
+      capabilities = null,
+      teleconsultPhoneNumber = "+919995558883"
```
Comment (msasikanth) [inline app/src/sharedTest/java/org/simple/clinic/TestData.kt:1076]: This can be an actual phone number, let's go with `+911111111111`

Comment (msasikanth) [inline app/src/sharedTest/java/org/simple/clinic/TestData.kt:1076]: Let's provide the option to override these values from `ongoingLoginEntry` method params or else it would be a hidden behaviour.

**Review by msasikanth — CHANGES_REQUESTED** [review]
PR looks good to me. Left couple of comments PTAL

File: app/src/test/java/org/simple/clinic/login/pin/LoginPinScreenLogicTest.kt [hunk 19]
```diff
@@ -128,7 +128,8 @@ class LoginPinScreenLogicTest {
         loggedInStatus = User.LoggedInStatus.OTP_REQUESTED,
         registrationFacilityUuid = registrationFacilityUuid,
         currentFacilityUuid = registrationFacilityUuid,
-        teleconsultPhoneNumber = teleconsultPhoneNumber
+        teleconsultPhoneNumber = teleconsultPhoneNumber,
+        capabilities = null
```
Comment (msasikanth) [inline app/src/test/java/org/simple/clinic/login/pin/LoginPinScreenLogicTest.kt:132]: You can do it in next PR, can we also pass capabilities here? Look at how we passed `teleconsultPhoneNumber`

**Review by msasikanth — APPROVED** [review]
LGTM
